### DISCUSSION TITLE
Idea StatiCa RCS API vs. RCS Controller

### DISCUSSION THREAD

**Discussion by kroovers** [idea-statica/ideastatica-public#436] [Help Q&A] [answered]
Hi Nathan,

During the “Learn Grasshopper” webinar I heard your announcement of a brand new RCS API, which leaves me with some questions.

Would you mind answering the following questions for me?

- What is the motivation for switching from the RCSController (old way of interacting with Idea RCS) to the RCS API?
- Does the API allow everything that was possible using the RCSController? Or better yet, does it enable new possibilities? If so, which?
- Does this API come with an expanded IOM support for RCS, adding some of the object definitions that were still missing, such as: 
  - Rebar group definitions by spacing
  - Layers of shear links definition
  - SetupValues that require tabular input (e.g. W_max)
- Will the RCSController remain functional in the coming versions, or will support end and should we switch to the API?
- Apart from the Python and C# sample codes on github, is there documentation available on the use of the different REST API end points?

Thanks!
Kelvin

**Comment by nathanlukeIDEA** [comment, marked as answer]
Hi @kroovers,

Some answers below:

-  What is the motivation for switching from the RCSController (old way of interacting with Idea RCS) to the RCS API?

_We basically could not maintain the RCSController, users had lots of issues with installing it and we could not manage it across different versions of IDEA StatiCA._ 

- Does the API allow everything that was possible using the RCSController? Or better yet, does it enable new possibilities? If so, which?

_Yes it should. Plus much much more. The primary additional functionality:_

-- _Ability to update a section's reinforcement directly through the API (using a defined .Nav file)._ 
-- _Change a sections reinforced cross-section directly through the API._
-- _Ability to update a section's loading directly through the API._
-- _Ability to update code settings directly through the API._
-- _Now we can reference library materials (using LoadFromLibrary switch instead of having to define all the properties manually in IOM._

- Does this API come with an expanded IOM support for RCS, adding some of the object definitions that were still missing, such as:
- Rebar group definitions by spacing
- Layers of shear links definition
- SetupValues that require tabular input (e.g. W_max)

_No additional IOM support (except the library material loading) has been added as a part of this. Regarding rebar, it can now be added via an API call. Let me know if this is useful in helping your issue or whether we still need further improvements in the IOM to enable you to achieve what you are after._ 

_Regarding setup values. This should be possible directly through the API and does not need to be set in IOM. You can test it using the RcsApiClient Testing App_  

![image](https://github.com/idea-statica/ideastatica-public/assets/93578548/a4947409-b1fe-4235-9d65-19de441c27a2)

- Will the RCSController remain functional in the coming versions, or will support end and should we switch to the API?
_It should remain functional, however, we will not support it anymore and would suggest upgrading to the new API._

- Apart from the Python and C# sample codes on github, is there documentation available on the use of the different REST API end points?

_We don't have the rest documentation currently up on our developer site. We can provide the swagger documentation to you if you like._

**Reply by kroovers** [reply]
Hi @nathanlukeIDEA,

Thanks for your reply, I'll give the new API a try and will get back to you if I get stuck.

And yes, if you could provide me the Swagger documentation so I can get insight in all the API end points, that would be very helpful!
Could you share with me the link to the swagger docs please?

Thanks for the help!

**Reply by nathanlukeIDEA** [reply]
Hi @kroovers, I have added the REST documentation to the developer docs. https://developer.ideastatica.com/docs/api/rcs/RCSAPI-oa2.html. 

The output of the documentation is not that pretty, a more useful way may be to download the source file and use the free swagger editor to view it. 

Please let us know if there are use cases that you would like to achieve or promote. We are looking to improve it. :)

**Reply by khameeteman** [reply]
Hi @nathanlukeIDEA, we are having a similar case as Kelvin.

Using Python we are building the IOM XML file, which we then insert into the RCSController to calculate the project. What would be the best way to convert this to the RCS API? We do not want to build up the entire model using the api, but stick to the XML file as starting point, is this possible?

**Reply by mpospisil** [reply]
Hi
What about to put basic templates from RCS to the API ? Users could set same parameters as in the desktop by API. It could solve this issue. It would be a parametric solution.

**Comment by kroovers** [comment]
Hi @nathanlukeIDEA 

I finally had the chance to look at the RCSApi to check whether it meets all out current needs in order to switch from RCSController to API. In general I'm happy with how it works, in principle it is indeed an improvement over the RCSController, so thanks for this upgrade! However, there are still some blocking points that I wanted to check with you:

_-- Ability to update a section's reinforcement directly through the API (using a defined .Nav file)._
We wish to define a Layer of Rebar by Spacing (for slabs, where n might be a floating point number). However, when exporting such a definition to .NAV it is exported as individual rebar (n is integer), thus when importing it back again the definition is changed (n = 2.4 -> n = 2). If this is a limitation of the .NAV format, then we would need a different way to define / update a layer of rebar by spacing. I've noticed in the past that it isn't supported using OpenModel either. The only workaround I ever found was to define this in the XML formatted version of the .idearcs file (available through the UI > SaveAs > *.xml or through the SingleCheckProject object). Do you know a solution that would work with the API?

Before export to .nav: 
<img width="676" alt="image" src="https://github.com/user-attachments/assets/cc8fd331-ba26-46ab-af9b-20fbc52a78b2">
After import of .nav: 
<img width="645" alt="image" src="https://github.com/user-attachments/assets/0eb8ff7a-1e81-4543-a58a-a8788ddfd140">

_-- Ability to update code settings directly through the API._
I have tried this for the NA_7_3_1_Wmax_Ec2_1_1 setup value, but couldn't get it to work. Would you happen to have a working sample?
I tried providing value as XML string, json string, and CI.Services.Setup.SetupTable_W_max_1992_1_1 object, all without effect.

```C#
var setup_W = new CI.Services.Setup.SetupTable_W_max_1992_1_1();
var xml = XElement.Parse("<Value><X0_XC1_RC>0.0001</X0_XC1_RC><XC2_XC3_RC>0.0002</XC2_XC3_RC><XD_XS_XF_RC>0.0003</XD_XS_XF_RC><X0_XC1_PC>0.0004</X0_XC1_PC><XC2_XC3_PC_DV>0.026</XC2_XC3_PC_DV><XC2_XC3_PC_CV>0.0002</XC2_XC3_PC_CV><XD_XS_XF_PCB_DV>0.025</XD_XS_XF_PCB_DV><XD_XS_XF_PCB_CV>0.0002</XD_XS_XF_PCB_CV><XC2_XC3_PCB_DB>true</XC2_XC3_PCB_DB><XC2_XC3_PCB_CB>true</XC2_XC3_PCB_CB><XD_XS_XF_PCB_DB>true</XD_XS_XF_PCB_DB><XD_XS_XF_PCB_CB>false</XD_XS_XF_PCB_CB></Value>");
setup_W.ReadXElement(xml);

var newSettings = new List<RcsSetting>()
{
     new RcsSetting()
     {
         Id=10, // NA_7_3_1_Wmax_Ec2_1_1
         Type="SetupValueEc", 
         //Value="<X0_XC1_RC>0.0001</X0_XC1_RC><XC2_XC3_RC>0.0002</XC2_XC3_RC><XD_XS_XF_RC>0.0003</XD_XS_XF_RC><X0_XC1_PC>0.0004</X0_XC1_PC><XC2_XC3_PC_DV>0.026</XC2_XC3_PC_DV><XC2_XC3_PC_CV>0.0002</XC2_XC3_PC_CV><XD_XS_XF_PCB_DV>0.025</XD_XS_XF_PCB_DV><XD_XS_XF_PCB_CV>0.0002</XD_XS_XF_PCB_CV><XC2_XC3_PCB_DB>true</XC2_XC3_PCB_DB><XC2_XC3_PCB_CB>true</XC2_XC3_PCB_CB><XD_XS_XF_PCB_DB>true</XD_XS_XF_PCB_DB><XD_XS_XF_PCB_CB>false</XD_XS_XF_PCB_CB>"
         //Value="{\"X0_XC1_RC\":0.0001, \"XC2_XC3_RC\":0.0002, \"XD_XS_XF_RC\":0.0003, \"X0_XC1_PC\":0.0004, \"XC2_XC3_PC_DV\":0.026, \"XC2_XC3_PC_CV\": 0.0002, \"XD_XS_XF_PCB_DV\": 0.025, \"XD_XS_XF_PCB_CV\":0.0002, \"XC2_XC3_PCB_DB\":true, \"XC2_XC3_PCB_CB\": true, \"XD_XS_XF_PCB_DB\": true, \"XD_XS_XF_PCB_CB\": false}"
         Value = setup_W
     },
};
var updateSuccess = await client.UpdateCodeSettings(newSettings, CancellationToken.None); // returns true, but settings aren't changed in the project file 
```


_-- Now we can reference library materials (using LoadFromLibrary switch instead of having to define all the properties manually in IOM._
Great that this works now! However, I notice that when enabling this switch, all properties are loaded from library and it's no longer possible to override some properties (assuming you want to load all values from library but only want to modify one or two). So currently you either have to define every property or can't define any property. Is there a middle way possible?

Thanks for the help!
